src/core_plugins/kibana/public/dashboard/actions/embeddables.js

```diff
@@ -32,6 +34,20 @@ export const embeddableRenderError = createAction('EMBEDDABLE_RENDER_ERROR',
   (panelId, error) => ({ panelId, error })
 );
 
+/**
+ * Converts the returned Embeddable object supplied by the embeddable handler into something we want to store on
+ * state.  We may need to store a cache of these somewhere so we can still have access to functionality.
+ * @param embeddable{Embeddable}
+ * @return embeddableState {EmbeddableState}
+ */
+function convertEmbeddableToState(embeddable) {
+  return {
+    title: embeddable.title,
+    editUrl: embeddable.editUrl,
```

Comment:
do we need to store renderComplete to state ?

Reply:
What else are you thinking?  Storing them on the state fits in with the whole redux theme and triggers the re-renders when this changes so anyone who needs this info can be updated if need be.

One idea I've had is to cache these Embeddables somewhere so the class objects themselves can be used - as it is you are only supposed to store primitive data in a redux tree.  We'll probably need to do that at some point anyway as the embeddables api grows.  So we could possibly store it in a cache and use it that way, but even then, I *think* we want it on state so the updates propagate.

@ redux folks - @chrisronline, @kjbekkelund, @w33ble, @bmcconaghy -- Have any of you run across the question of how or where to store class objects in the redux world?  In dashboard, I have pluggable EmbeddableHandlers, and Embeddables, which are classes.  If you have class objects, do you cache them outside of the store?  Maybe taking pieces that are relevant to the store (like I am here) and storing it in the redux tree?  But then you run the risk of it getting out of sync?

I think in the redux world you aren't really supposed to have class objects - you have state and you have functions which manipulate that state.  Maybe my pluggable set up could be improved? @w33ble - in canvas you guys have pluggable concepts.  Have you dealt with this question/issue at all, or do you just make sure your state and your functionality is always separate?

Reply:
Interesting. I definitely agree that we should not store class objects within the redux store, unless there is some clear serialization path that will work with redux but that might be more effort than it's worth. I think the approach you have here is what I'd do too.

> But then you run the risk of it getting out of sync?

Can you explain more how this might happen?

Reply:
> Can you explain more how this might happen?

hmmm, here's a hypothetical scenario:
1. EmbeddableHandler returns Embeddable with title `xyz`.  We store `xyz` for the title in redux on the state tree.
2. User interacts directly with the embeddable, which changes `xyz` title on the embeddable to `abc`.  We don't know anything about it because it happened directly on the class object, so we will continue to show `xyz`.

Maybe this would just be "wrong" so unsupported.

If we always grabbed the title directly from the Embeddable, we'd always get the most recent value, but then we are operating outside of the redux world and there are other complications.

I guess I'm just not sure of the best practices when dealing with redux and a pluggable system.  Maybe we just need to be strict about the requirement to always change state through the communication layer.

Reply:
Possibly crazy thought, have you explored configuring the embeddable with a redux action (or just a callback) that should be called when its internal state changes?

I think we want to restrict the data source to a single place but we can normalize the state across other areas if necessary. We'd probably just need a mechanism to keep them up to date. 

Another potentially crazy thought - have you explored storing the embeddable state _only_ within a redux tree? I'm not sure what that might look like for the non-react consumers (assuming you have these) but redux should work without react so it might make more sense to put everything there. This might be a bad idea too as it would make some of these systems to dependent on redux (but maybe that's not so bad).

Reply:
>  have you explored storing the embeddable state only within a redux tree? 

I think this could be dangerous because then the creator of the embeddable might directly update state, not realizing that it's part of our redux tree.  We probably want to specifically separate out the two to avoid an Embeddable updating state in our redux tree directly.

> This might be a bad idea too as it would make some of these systems to dependent on redux (but maybe that's not so bad).

I'm pretty confident I want the embeddables to be framework agnostic.  Esp since our current vis and saved searches are in angular.

I also know that the new platform has a pluggable system, and @epixa mentioned to me that the new platform solves a lot of the problems the embeddable plugin system tries to solve, so maybe the best flow will become clearer once that's in place.

Reply:
> I think this could be dangerous because then the creator of the embeddable might directly update state, not realizing that it's part of our redux tree. We probably want to specifically separate out the two to avoid an Embeddable updating state in our redux tree directly.

Why does that concern you exactly? It sounds like the state of the embeddable is directly tied  to the redux state so changes in the former should immediately propagate to the latter and the redux tree should handle that change appropriately (re-renders, etc). Or are there scenarios where the embeddable state changes, but the redux tree shouldn't update because of it?

Reply:
I think synchronization between the non-redux interface of embeddable and the internal redux state for dashboard is the right move here.  This is exactly how basically *all* interactions between systems will occur in new platform world.

It sort of sounds like each embeddable should instead be consumed by the dashboard as an observable, so you can subscribe to changes in embeddable and dispatch the appropriate action to update the internal state.

Some rough pseudocode to illustrate:

```js
// wire up the incoming embeddables when you bootstrap the redux store
const store = createStore();

// every new value is either a brand new thing to embed or an update to an existing thing
embeddablesObservable.subscribe(embeddable => {
  store.dispatch(syncEmbeddableAction(embeddable));
});
```

That example assumes the difference between "new thing on this dashboard" and "changing the local details of an existing thing on the dashboard" can be handled behind the scenes in `syncEmbeddableAction`, but you could certainly wire up observables in a different way that is more explicit about new vs existing.

Reply:
hmm, interesting, that sounds like an approach worth exploring @epixa. I'll have to think a bit more about how to formalize that approach.

@chrisronline re:
> Why does that concern you exactly? It sounds like the state of the embeddable is directly tied to the redux state so changes in the former should immediately propagate to the latter and the redux tree should handle that change appropriately (re-renders, etc). Or are there scenarios where the embeddable state changes, but the redux tree shouldn't update because of it?

Maybe we are not quite on the same page, but if the redux tree contains a reference to state that the embeddable also holds, and the embeddable updates that state directly (e.g. `embeddable.title = "abc"`) then this will break redux because you are never supposed to mutate the redux tree directly, which is why instead a copy is always returned from the reducers.  If it's mutated directly comparisons will break next time around for when to trigger re-renders.

Reply:
Ah, I didn't mean that exactly.

More like:
```
class EmbeddableHandler {
  setChangeHandler(handler) {
    this.changeHandler = handler;
  }
  notifyOfChange() {
    this.changeHandler();
  }
}
```

then in your action:
```
export const embeddableUpdated = createAction('EMBEDDABLE_UPDATED');

export function renderEmbeddable(embeddableHandler, panelElement, panel, containerApi) {
...
return embeddableHandler.render(panelElement, panel, containerApi)
      .then(embeddable => {
        const embeddableState = convertEmbeddableToState(embeddable);

        // maybe a better place for this, but just go with it for now :)
        embeddable.setChangeHandler(() => dispatch(embeddableUpdated(...)));

        embeddable.renderComplete.then(() => dispatch(embeddableRenderComplete(panel.panelIndex)));

```

Then, you subscribe to changes on the class object and use those changes to re-update the redux state tree through the expected action/reducer mechanisms.